ackermann_steering_controller/test/test_ackermann_steering_controller.hpp

```diff
@@ -126,9 +126,20 @@ class AckermannSteeringControllerFixture : public ::testing::Test
       "/test_ackermann_steering_controller/reference", rclcpp::SystemDefaultsQoS());
   }
 
-  static void TearDownTestCase() {}
+  void TearDown()
+  {
+    try
+    {
+      controller_->get_node()->shutdown();
+    }
+    catch (...)
+    {
+      // ignore case where node is not initialized
+    }
```

Comment:
```suggestion
    if(controller_->get_node().get_state() != lifecycle_msgs::msg::State::PRIMARY_STATE_UNKNOWN)
    {
      controller_->get_node()->shutdown();
    }
```

I think it is better to have it like this no?. trying and catching everything might hide some issues we might add in future. What's your opinion on this @christophfroehlich ?. If you think it's fine, I'm okay with it

Reply:
The following code
```c++
  if(range_broadcaster_->get_node()->get_current_state().id() != lifecycle_msgs::msg::State::PRIMARY_STATE_UNKNOWN)
  {
    range_broadcaster_->get_node()->shutdown();
  }
```

still throws
```

2: [ RUN      ] RangeSensorBroadcasterTest.Initialize_RangeBroadcaster_Exception
2: unknown file: Failure
2: C++ exception with description "Lifecycle node hasn't been initialized yet!" thrown in TearDown().
```

here

https://github.com/ros-controls/ros2_control/blob/d8a21c4e5effa2ac3b75a3dffbc6b48471e18a1f/controller_interface/src/controller_interface_base.cpp#L205-L212

I don't see any other possibility with the controller_interface API.

Reply:
we could only catch `std::runtime_error` and rethrow others. 
Or we change the upstream API to check if the node is initialized or add a shutdown method, which itself checks if the node is initialized.

Reply:
Ahh I didn't know that this exception is in the controller interface base. If this is the case, we can add an upstream API to check if the controller is initialized or may be in the destructor of the base class, call shutdown if the node is valid and also the state makes sense. What do you think?

Reply:
we never trigger the shutdown transition in the controller_manager, maybe this is a bug we should solve?
Currently we only have virtual default destructors of the `ControllerInterfaceBase` and `ControllerInterface` class, where would you add that?

Reply:
I would say in the destructor of the controller_interface_base. As that's the common one for both normal and chained controllers